core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala

```diff
@@ -166,9 +172,12 @@ private[spark] class MetricsSystem private (
   }
 
   def removeSource(source: Source) {
-    sources -= source
-    val regName = buildRegistryName(source)
-    registry.removeMatching((name: String, _: Metric) => name.startsWith(regName))
+    val index = sources.indexOf(source)
+    if (index != -1) {
```

Comment:
If there use `sources.contains(source)`, it looks more graceful.

Reply:
use index can avoid navigating sources twice.

Reply:
using index can avoid navigating sources twice.

Reply:
I know it, now.